### PR TITLE
Fix syscall deprecation warning on macOS >= 10.12

### LINKED ISSUE #185: OSX 10.12 compilation warning: syscall is deprecated
The relevant snippets are below.

```
/Users/pwnall/workspace/glog/src/raw_logging.cc:153:3: warning: 'syscall' is deprecated: first deprecated in
      macOS 10.12 - syscall(2) is unsupported; please switch to a supported interface. For SYS_kdebug_trace use
      kdebug_signpost(). [-Wdeprecated-declarations]
  safe_write(STDERR_FILENO, buffer, strlen(buffer));
  ^
/Users/pwnall/workspace/glog/src/raw_logging.cc:63:34: note: expanded from macro 'safe_write'
# define safe_write(fd, s, len)  syscall(SYS_write, fd, s, len)
                                 ^
/usr/include/unistd.h:733:6: note: 'syscall' has been explicitly marked deprecated here
int      syscall(int, ...);
         ^
1 warning generated.
```

```
/Users/pwnall/workspace/glog/src/utilities.cc:254:17: warning: 'syscall' is deprecated: first deprecated in
      macOS 10.12 - syscall(2) is unsupported; please switch to a supported interface. For SYS_kdebug_trace use
      kdebug_signpost(). [-Wdeprecated-declarations]
    pid_t tid = syscall(__NR_gettid);
                ^
/usr/include/unistd.h:733:6: note: 'syscall' has been explicitly marked deprecated here
int      syscall(int, ...);
         ^
1 warning generated.
```

### PR DESCRIPTION
**Pull request by z-aki**
Fix #185

### CODE REVIEW

**Comment by z-aki** [comment]
@googlebot I signed it!

File: src/utilities.cc [hunk 1]
```diff
@@ -248,7 +248,13 @@ pid_t GetTID() {
 #endif
   static bool lacks_gettid = false;
   if (!lacks_gettid) {
+#if (defined OS_MACOSX && MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_12)
```
Comment (sergiud) [inline src/utilities.cc:251]:
```suggestion
#if (defined(OS_MACOSX) && MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_12)
```

File: src/raw_logging.cc [hunk 2]
```diff
@@ -59,7 +59,8 @@
 # include <unistd.h>
 #endif
 
-#if defined(HAVE_SYSCALL_H) || defined(HAVE_SYS_SYSCALL_H)
+#if defined(HAVE_SYSCALL_H) || defined(HAVE_SYS_SYSCALL_H) && \
+  !(defined OS_MACOSX && MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_12)
```
Comment (sergiud) [inline src/raw_logging.cc:63]:
```suggestion
  !(defined(OS_MACOSX) && MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_12)
```

**Review by sergiud — CHANGES_REQUESTED** [review]
Currently, it is not possible to test these changes as a corresponding Mac OS X Github workflow is missing.

**Comment by z-aki** [comment]
Err is that failure my fault ? macOS guards shouldn't affect mingw..

**Comment by sergiud** [comment]
The failure is very likely caused by this branch.

**Comment by z-aki** [comment]
Okay, so far I was building glog as part of a downstream project. 
Will check separately now.

**Comment by z-aki** [comment]
Okay it was operator precedence. Fixed now.

**Comment by sergiud** [comment]
Thanks. Could you please rebase onto current master? CI now supports macOS.

**Comment by sergiud** [comment]
macOS builds still issue a warning, e.g.,
```
/Users/runner/work/glog/glog/src/raw_logging.cc:140:3: warning: 'syscall' is deprecated: first deprecated in macOS 10.12 - syscall(2) is unsupported; please switch to a supported interface
```

**Comment by sergiud** [comment]
Warning still present.

**Comment by z-aki** [comment]
Shouldn't the glog project be defining  `MAC_OS_X_VERSION_MAX_ALLOWED` since you know which OSes it will be build on? 
If you can't I can just remove the macOS version stuff and make it dependent on OS only.

**Comment by sergiud** [comment]
The library doesn't care about `MAC_OS_X_VERSION_MAX_ALLOWED`, hence it does need to be defined.

Can't you just check the presence of `pthread_threadid_np` in CMake using `check_function_exists` and use the function if it is available instead of relying on defines?

File: src/config.h.cmake.in [hunk 3]
```diff
@@ -222,4 +222,6 @@
 
 #endif
 
+#cmakedefine HAVE_PTHREAD_THREADID_NP ${HAVE_PTHREAD_THREADID_NP}
```
Comment (sergiud) [inline src/config.h.cmake.in:225]: Please add a comment describing the define for consistency.

File: CMakeLists.txt [hunk 4]
```diff
@@ -140,6 +140,8 @@ check_cxx_compiler_flag (-Wdeprecated HAVE_NO_DEPRECATED)
 check_cxx_compiler_flag (-Wunnamed-type-template-args
     HAVE_NO_UNNAMED_TYPE_TEMPLATE_ARGS)
 
+check_symbol_exists (pthread_threadid_np "pthread.h" HAVE_PTHREAD_THREADID_NP)
```
Comment (sergiud) [inline CMakeLists.txt]:
```suggestion
cmake_push_check_state (RESET)
set (CMAKE_REQUIRED_LIBRARIES Threads::Threads)
check_cxx_symbol_exists (pthread_threadid_np pthread.h HAVE_PTHREAD_THREADID_NP)
cmake_pop_check_state ()
```

**Review by sergiud — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 5]
```diff
@@ -259,6 +260,13 @@ void foo(std::nullptr_t) {}
 int main(void) { foo(nullptr); }
 " HAVE_CXX11_NULLPTR_T)
 
+check_cxx_source_compiles ("
```
Comment (sergiud) [inline CMakeLists.txt:263]: This is unnecessary.

**Comment by z-aki** [comment]
nope warning's still there..
It's sometimes useful to clear cache when dealing with cmake variables.
Also, in one of the older attempts, I just set the code to 
```
#if (defined(OS_MACOSX))
    uint64_t tid64;
    pthread_threadid_np(NULL, &tid64);
    pid_t tid = (pid_t)tid64;
...
```
and it built fine. 
The function was introduced in macOS 10.6 so seems safe to use now.

File: src/utilities.cc [hunk 6]
```diff
@@ -248,7 +248,13 @@ pid_t GetTID() {
 #endif
   static bool lacks_gettid = false;
   if (!lacks_gettid) {
+#if (defined(OS_MACOSX) && defined(HAVE_PTHREAD_THREADID_NP))
+    uint64_t tid64;
+    pthread_threadid_np(NULL, &tid64);
```
Comment (sergiud) [inline src/utilities.cc:253]: You should probably check the return value and assign -1 to `tid` in case of an error.

**Review by sergiud — CHANGES_REQUESTED** [review]
I have just one minor remark. You can then squash the commits.

**Review by sergiud — APPROVED** [review]

**Comment by sergiud** [comment]
Great. Thanks!

**Pull request merged**